### PR TITLE
Switch build-info and build images to run on public runners

### PR DESCRIPTION
**Pull request by potiuk**
We used to have low number of Public runners on GitHub actions and Apache Software Foundation projects were competing for them, that's why we switched `build-images` workflow to run on self-hosted runners, and we had to add workflow-embedded list of committers to make sure that build-info also uses self-hosted runners.

Recently however the ASF has 900 runners and if anything, it's the public runners that are in abundance, while self-hosted runners are limited, which sometimes leads to "Wait for CI images" jobs timeout while waiting for self-hosted runners to build the image.

Those "wait-for-ci-images" jobs do almost nothing, just waiting, so using self-hosted runners for those is not needed, we can easily switch those to public runners as well.

This PR does the following changes:

* build-info jobs always run on public runners
* wait-for-ci-images job are always run on public runners
* all the other jobs run on either self-hosted or public runners depending who the actor is (committers - run on self-hosted, non-committers, run on public runners)

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -65,64 +65,7 @@ jobs:
     # to find the list of authors and replace them, so any changes to the
     # formatting of the contains(fromJSON()) structure below will need to be
     # reflected in that workflow too.
-    runs-on: >-
```
Comment (potiuk) [inline .github/workflows/ci.yml:68]: Finally we will be able to get rid of this list in yaml (@ashb - ci-infra instruction fix will follow once we merge it).

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -525,7 +468,7 @@ jobs:
   wait-for-ci-images:
     timeout-minutes: 120
     name: "Wait for CI images"
-    runs-on: "${{needs.build-info.outputs.runs-on}}"
```
Comment (potiuk) [inline .github/workflows/ci.yml:528]: I think we will save quite a lot of "self-hosted" time for committer builds this way.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -548,13 +491,6 @@ jobs:
         env:
           PYTHON_VERSIONS: ${{ needs.build-info.outputs.python-versions-list-as-string }}
           DEBUG_RESOURCES: ${{needs.build-info.outputs.debug-resources}}
-      - name: "Tests Pytest collection"
```
Comment (potiuk) [inline .github/workflows/ci.yml:551]: This wil save ~ 4 minutes of "feedback time" as this test-pytest-collection is also repeated in each test job.

File: .github/workflows/build-images.yml [hunk 4]
```diff
@@ -66,6 +65,7 @@ jobs:
       cache-directive: ${{ steps.selective-checks.outputs.cache-directive }}
       default-branch: ${{ steps.selective-checks.outputs.default-branch }}
       default-constraints-branch: ${{ steps.selective-checks.outputs.default-constraints-branch }}
+      runs-on: ${{ steps.selective-checks.outputs.runs-on }}
```
Comment (potiuk) [inline .github/workflows/build-images.yml:68]: This should also save quite a lot of "self-hosted" time -  building images will now be done on public runners for non-commiter runs.

File: dev/breeze/src/airflow_breeze/utils/selective_checks.py [hunk 5]
```diff
@@ -765,3 +777,29 @@ def affected_providers_list_as_string(self) -> str | None:
...
+                    "\n[error]ERROR: Needs GitHub token to determine if it runs on self-hosted runner."
+                )
+                get_console().print(
+                    "\n[info]You can generate run at "
+                    "https://github.com/settings/tokens/new?description=Read%20Org&scopes=read:org\n"
+                )
+                sys.exit(1)
+            from github import Github
+
+            gh = Github(self._github_token)
```
Comment (potiuk) [inline dev/breeze/src/airflow_breeze/utils/selective_checks.py]: Committer list is now retrieved dynamically. We do not need to hard-code it at all.

Comment (potiuk) [reply dev/breeze/src/airflow_breeze/utils/selective_checks.py]: Unfortunately, we have no permissions in the actions to read the list :(, I needed to re-hardcode it.

**Review by jedcunningham — APPROVED** [review]

**Review by hussein-awala — APPROVED** [review]

**Comment by potiuk** [comment]
Let's see. I will merge and see how things will play out :) 

Living on the edge ;)

**Pull request merged**